### ISSUE TITLE
[Bug Report] Latest update to moreh_adamw has issue 

### ISSUE THREAD

**Issue by rfurko-tt** [tenstorrent/tt-metal#14186]
**Describe the bug**
After updating to latest `tt-metal` main branch, we see that **training loss goes down way slower**. After turning off program cache, training loss goes as expected. We expect that bug is located in latest update to `moreh_adamw` cache functions. One of the ideas, that we aren't properly passing `step` now. 

**Expected behavior**
Speed of convergence should not be affected by program caching. 

**Please complete the following environment information:**
- OS: Ubuntu 20.04
- Version of software 05ae26fa027ff80e45614a9d324acc0d36940718

**Comment by o2buzzle** [comment]
That is odd to say the least.

I would not expect that to happen either since the program cache *should* be faster. It being slower would suggests runtime argument replacement is taking longer than recompiling from scratch which makes little sense to me

**Comment by mrshaw01** [comment]
Hi Roman,
'Goes way slower' here means it's slower when running the same number of steps. Or: with n steps, it takes the same amount of time to run, but the loss is larger when the program cache is turned on compared to when it's turned off.

**Comment by dmakoviichuk-tt** [comment]
@mrshaw01  we mean training loss.
Feels like step is not updated.

**Comment by rfurko-tt** [comment]
Hi everyone,
Sorry for confusion. Loss goes down way slower.

**Comment by dmakoviichuk-tt** [comment]
But if talk about cache. Somehow I don't see significant perf difference in cached vs non-cached :)

**Comment by rfurko-tt** [comment]
Before cache fix, we get loss equal to 3.0 at iteration 8.  
After cache fix,  we get to this value only after 90 iterations.

**Comment by mrshaw01** [comment]
We’ve got it. o2buzzle will debug this issue.
Thank you for the information!

**Comment by o2buzzle** [comment]
@rfurko-tt That would suggest an issue with `override_runtime_arguments` not overriding kernel runtime values, but so far I couldn't find anything obviously wrong there yet, so it's likely something else deeper in the framework.

`DPRINT` inside the kernel shows that both are updating as expected. 
![image](https://github.com/user-attachments/assets/9084700c-24a0-4706-a468-49efe53c72c6)

**Comment by dmakoviichuk-tt** [comment]
@o2buzzle  that was our idea because it kinda trained but like with 100x smaller lr.

**Comment by o2buzzle** [comment]
@dmakoviichuk-tt Would you mind providing some sample code that can reproduce the issue for reference? As it is I am not able to find anything that would stands out as buggy and the unit tests seems to be agreeing with that.

**Comment by dmakoviichuk-tt** [comment]
Hi @o2buzzle  unfortunately it is very hard to show this code right now.
What we found:
https://github.com/tenstorrent/tt-metal/commit/15338d430ba1448939911008e50f31ec7189f67c this commit works for us.
But when you introduced caching for learning rate it stopped working.

**Comment by dmakoviichuk-tt** [comment]
https://github.com/tenstorrent/tt-metal/pull/14243 here is the fix. please approve

**Comment by o2buzzle** [comment]
@mrshaw01 dude asked me to do so, fml

**Comment by mrshaw01** [comment]
Both approaches should work well.
The 'manual addition' approach has already been used in some operations, so I suggested it for consistency. However, it turned out to be buggy.
But nevermind, the current approach is OK.

**Comment by dmakoviichuk-tt** [comment]
@mrshaw01 would be nice to find why it is buggy in fuutre. Is it hash or missed parameters?

**Comment by o2buzzle** [comment]
I find the odd thing here being that the kernel itself is reporting (through `DPRINT`) that the value has correctly changed, if I wrote an unit test to modify it, so whatever has been happening, it must have somehow hit the cache but managed to bypass runtime argument replacements for this to happen. Maybe some quirk in the caching mechanism?

I originally study and re-implemented the hashing code from the example at `group_attn_matmul_device_operation.cpp`, and that seems to work fine.
